---
title: "Mouse Whole Cortex and Hippocampus 10x dataset"
author: "Lara Ianov, Ph.D."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

This is a data wrangling example from the Mouse Whole Cortex and Hippocampus 10x dataset.

Recently, a UAB collaborator, Dr. Farah Lubin and her lab, asked U-BDS to summarize a subset of the the Allen Brain Map data derived from the Mouse Whole Cortex and Hippocampus 10x dataset. A driver for this request was due to the limitation in the dataset that is provided directly from the website. For example, while users can choose to view any gene per celltype https://celltypes.brain-map.org/rnaseq/mouse_ctx-hpf_10x?selectedVisualization=Heatmap&colorByFeature=Cell+Type&colorByFeatureValue=Gad1, it's difficult to summarize the data based on cell classes. In this example, we illustrate how you can start from the data available at the Allen Brain Atlas website, to interpret the data in more detail.

__The gene list in this example differs from the original request by Dr. Lubin, and user can choose change it to any gene list of interest.__

* Input data from: https://portal.brain-map.org/atlases-and-data/rnaseq/mouse-whole-cortex-and-hippocampus-10x
* The same version of the data used in this guide can be acquired at: https://uab.box.com/s/2ye54fsiue7sq97nmvbh46sb806ya6i9
* Publication: https://pubmed.ncbi.nlm.nih.gov/34004146/
* __Goal:__ to acquire summarized values across classes of cells label (e.g.: GABAergic, glutamatergic etc.) rather than viewing them from the app which only provides values split by `subclass_label`/`cell_type_alias_label`.

To stay consistent with what the user sees in the app, the input data is the trimmed mean expression values which are available from the download pages cited above.

In addition, the metadata file is also needed.

## Information about available data

In the README from the data download page, we find the following information about the `medians.csv` which was generated in a similar manner as the trimmed means file.

`medians.csv` file

>Medians (medians.csv)
	A table of median expression values for each gene (rows) in each cluster (columns).  Medians are calculated by first normalizing gene expression as follows: norm_data = log2(CPM(exons+introns)), and then calculating the medians independently for each gene and each cluster.
	The first row lists the cluster name (cluster_label), which matches the cell type alias shown in the Transcriptomic Explorer.
	The first column lists the unique gene identifier (gene), which in most cases is the gene symbol.

The `medians.csv` ("Gene Expression by Cluster, median") is briefly described from the data download page with:

>Gene expression aggregated per cluster, calculated as medians.

Similarly the trimmed means values ("Gene Expression by Cluster, trimmed means"), is described as:

>Gene expression aggregated per cell type, calculated as trimmed means (25%-75%).

The metadata description is:

>Metadata about each cell. Each entry has an _order, _label, and _color column.


# Metadata

Since the trimmed mean expression values are readily available, we will summarize this by metadata available in the `metadata.csv`.

```{r load packages, message=FALSE, warning=FALSE}
library(ggplot2)
library(dplyr)
```

The files needed were downloaded and placed into the `./data` sub-directory from this project which follows the general standard practice for R projects (see more at: https://swcarpentry.github.io/r-novice-gapminder/02-project-intro/index.html).

```{r import data}
trimmed_mean_expression <- read.csv("./data/Allen_Brain_Map_mouse_sc_data/trimmed_means.csv", check.names = FALSE, row.names = 1)
metadata <- read.csv("./data/Allen_Brain_Map_mouse_sc_data/metadata.csv")
```

```{r}
# If unfamiliar with dataset, please look over the metadata in more detail, View(metadata), for better understanding of all the information present here

# but here is a quick overview of some key features

str(metadata)
```


In particular columns named `class_label` and `cell_type_designation_label` contain cell classification which is of high interest to the main goal.

```{r}
unique(metadata$class_label)

head(unique(metadata$cluster_label)) # same as metadata$cell_type_alias_label

unique(metadata$subclass_label)
```

```{r, include = FALSE}
colnames(trimmed_mean_expression)
```

Thus, the column names from the trimmed means files matches values in `metadata$cluster_label`:

```{r}
identical(sort(unique(metadata$cluster_label)),
          sort(unique(colnames(trimmed_mean_expression)))
          )
```

Now, since the variables which links the two input files have been identified, the data can be merged and summarized.

## Filter metadata

First, filter the metadata to only keep the variables of interest highlighted above for simplicity. Do not keep `sample_name` since the trimmed values are summarized per cluster rather then single-cells. Further, since the original metadata has all information per cell, further filter this output to keep unique rows.

If new metadata is needed in the future, modify the following to keep the needed data.

```{r}
metadata %>%
    select(class_label, cluster_label, subclass_label)  %>%
    unique(.) %>%
    {.} -> metadata

```

# Reshape trimmed means

Transpose the trimmed means data.frame to have gene names as column names and add a new column which contains the `cluster_label` information (which are transposed as row names). This will allow us to merge the metadata information into a new data.frame and summarize and plot the data grouped by any variable that has been retained to this point.

```{r}
trimmed_mean_expression_transposed <- as.data.frame(t(trimmed_mean_expression))

trimmed_mean_expression_transposed$cluster_label <- rownames(trimmed_mean_expression_transposed)
```

Merge the two data.frames to have all needed data in a single data.frame and save this new data. Note there is a single row that contains values (trimmed means), but no cluster information. We will keep this here in this version, but could be optionally removed.

```{r}
trimmed_mean_with_metadata <- merge(metadata, trimmed_mean_expression_transposed, by = "cluster_label")

dir.create("results", showWarnings = FALSE)

write.csv(trimmed_mean_with_metadata, file = "./results/trimmed_means_with_selected_metadata.csv", row.names = FALSE)

# transposed version if that's preferred...

write.csv(t(trimmed_mean_with_metadata), file = "./results/trimmed_means_with_selected_metadata_transposed.csv", row.names = TRUE)

```


# Example plots and data summary

For plotting and summarizing this data, all that is needed is the data present in the `trimmed_mean_with_metadata` object created in this workflow (or the file saved as  `./results/trimmed_means_with_selected_metadata.csv`).

The following are just some examples of what we may now do with this new file/data:

## Violin plots for GABAergic class

The goal here would be to generate data split by `subclass_label` for only `GABAergic` class. This should match the same values we see in the heatmaps available online but gives us the flexibility of plotting is differently and also allows us to verify that our data wrangling above was properly performed.

To perform this:

* `filter` the data to contain the feature of interest (e.g.: `class_label` or `subclass_label`) 
* Plot violin plots for a gene of interest

For the following example, we will plot data for the gene `Gad1` for `class_label == GABAergic`

### Example of filtering

For a quick overview of data frame manipulation, please see https://swcarpentry.github.io/r-novice-gapminder/13-dplyr/index.html

```{r}
feature_to_filter_for <- "GABAergic"

trimmed_mean_with_metadata %>%
    filter(class_label == feature_to_filter_for) %>%
    {.} -> trimmed_mean_with_metadata_filtered

```

We now have the trimmed means values for GABAergic cluster in `trimmed_mean_with_metadata_filtered`. 

Or we can plot these values colored by `subclass_label`

```{r}

# Gad1
ggplot(trimmed_mean_with_metadata_filtered, 
       mapping = aes(x = subclass_label, y = Gad1, fill = subclass_label)) + 
    geom_violin() + geom_point() + theme_bw() + 
    ggtitle(paste0("Trimmed means values for data filtered for ", feature_to_filter_for))

```

## Summarizing data

Aside from generating figures, this data can be summarized in any form the user would like. To the original goal, we can compute the `median` for example, using the dataset which we have already pre-filtered (e.g. using `GABAergic` filtered data).

For example the median for `Gad1` across `GABAergic` class can be computed by:

```{r}
median(trimmed_mean_with_metadata_filtered$Gad1)
```

To summarize this across a list of genes, you may do the following:

1. Import a list of genes of interest, which in this case is saved to a file named `genes_of_interest.csv`

```{r}
genes_of_interest <- read.csv(file =  "./data/Allen_Brain_Map_mouse_sc_data/genes_of_interest.csv", header = TRUE)
```

2. Summarize across all genes:

There are several ways of accomplishing this type of goal. A common method is using `group_by` and `summarize` functions from `dplyr`. An overview of this can be seen at https://swcarpentry.github.io/r-novice-gapminder/13-dplyr/index.html

Here, we take another approach by working with the already filtered data `trimmed_mean_with_metadata_filtered` to acquire the medians and other stats of each gene across all `GABAergic` entries:


```{r}
trimmed_mean_with_metadata_filtered %>%
    select(genes_of_interest$genes_of_interest) %>%
    {.}-> trimmed_mean_with_metadata_filtered_by_gene


summary_data <- as.data.frame(apply(trimmed_mean_with_metadata_filtered_by_gene,2,summary))

#save

write.csv(summary_data, file = paste0("./results/summary_by_gene_of_interest_", feature_to_filter_for, ".csv"))

```


```{r, echo = FALSE}
DT::datatable(t(summary_data),
              options = list(
                  scrollX = TRUE
              ))
```

# `sessionInfo`

```{r}
sessionInfo()
```

