---
title: "nf-core tips"
author: "Austyn Trull and Lara Ianov, Ph.D."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# General Notes
* Use at minimum 3 samples when running a pipeline (even for minimal testing purposes). Some pipelines will not be able to complete successfully when running only a single sample.
* Although not a requirement, we advise to not use the iGenomes option suggested in a number of pipelines due to the lack of information on versioning linked to each assembly which may compromise reproducibility to a small degree. Instead pass on a reference (genome, and annotation file(s)) from a source and version that is suitable to your needs.

# Executor mode for Cheaha profile

* A PR to the nf-core/configs has been opened to add usage of executor mode for UAB's HPC as the Cheaha profile: https://github.com/nf-core/configs/pull/322 ; once approved users may implement this feature as described in the documentation linked to this profile. Any question or suggestions for this profile should be addressed to Austyn Trull and Dr. Lara Ianov. The addition of this profile has gone through the needed review steps with Research Computing prior to public submission of the PR.

We recommend to use this profile when processing multiple samples in the pipeline as it will aid in parallel processing. If your run only contains a small amount of samples, standard `local` mode (to a compute node) is usually sufficient.

# Pipeline-specific Notes
## nf-core/rnaseq

Although a user choice, we generally recommend to enable Salmon's flags `--seqBias --gcBias`. In version 3.4, enabling them can be done via a `custom.config`, by including the following (enabling the flags in the salmon run with star inputs, and with the salmon run where quasi-mapping was performed by salmon):

```
params {
  modules {
    'star_salmon_quant' {
      args = '--seqBias --gcBias'
    }
    'salmon_quant' {
      args = '--seqBias --gcBias'
    }
  }
}
```

Note that the syntax linked to custom parameters may differ from pipeline to pipeline or even versions. We recommend to following similar syntax to what is present in the `modules.config` of the pipeline of interest such as [this one for v.3.4](https://github.com/nf-core/rnaseq/blob/3.4/conf/modules.config). If you have question about this, please feel free to ask a question linked to it during our data science office hours.

### Version specific notes

* __3.5__: There has been an issue found with this version linked to the DESeq2 QC step. We have described a solution around with the description of the bug which can be found at: https://github.com/nf-core/rnaseq/issues/754
    * Alternatively, __3.4__ does not have this bug and we have not found any other issues.

### Tips linked to reference
* The `transcript_fasta` is not a required parameter by the pipeline. If this is not provided by the user, it will be generated by the process called `RSEM_PREPAREREFERENCE_TRANSCRIPTS` (which may be renamed in the near-future as this process is executed whenever the user does not provide a transcriptome, regardless of the choice of `--aligner`). For more information, reference [here](https://nfcore.slack.com/archives/CE8SSJV3N/p1641824929069500)
* __If the user provides a file to__ `transcript_fasta` which is used at the Salmon steps, it should be a transcriptome file generated from `gffread` as opposed to one downloaded from GENCODE or Ensembl. Issues linked to Ensembl transcriptomes are derieved from the fact that the transcriptome files (even after coding and non-coding RNA concatenation) have less transcripts than the GTF file. The issues linked to GENCODE files are linked to an issue in the pipeline which prevents it from correctly processing the off-the-shelf transcript fasta due to the sequence name format (despite enabling the `--gencode` parameter). For more information, reference [here](https://nfcore.slack.com/archives/CE8SSJV3N/p1642697218225700) and [here](https://nfcore.slack.com/archives/CE8SSJV3N/p1642699049230500)
    * It should be noted the current recommended way to use reference files generally is to provide the `--fasta` and `--gtf` along with `--save_reference` the first time you run the pipeline and then let it generate all of the downstream artifacts like the transcriptome and indices to be recycled thereafter via including them in them as params in the yml. This is the way recommended by the authors per the nf-core/rnaseq slack channel. For more information, reference [here](https://nfcore.slack.com/archives/CE8SSJV3N/p1642698848229000)

`gffread` example:

```
samtools faidx Mus_musculus.GRCm38.dna.primary_assembly.fa

gffread -w Mus_musculus.GRCm38_GTF_matched_transcripts.fa -g ./Mus_musculus.GRCm38.dna.primary_assembly.fa Mus_musculus.GRCm38.102.gtf
```

From the example above, the newly generated transcriptome `Mus_musculus.GRCm38_GTF_matched_transcripts.fa` would be passed on to the pipeline.

We recommend to creata a conda environment with both dependencies and record the versions.

`gffread` documentation can be found from: http://ccb.jhu.edu/software/stringtie/gff.shtml#gffread

* An issue has been opened with the nf-core team to add a warning regarding the points above linked to `transcript_fasta`: https://github.com/nf-core/rnaseq/issues/753

## nf-core/nanoseq
* The version of nanoplot that is used by this pipeline (1.38) causes the pipeline to crash. Per the author, the issue is with a package that is used to convert html files into png files. However, nanoseq expects there to be png files which causes crash. In order to get around this issue, you can create a custom config and override the container that nanoseq would pull down for Nanoplot. Values specified in configs passed via the commandline take priority over the defaults that are specified in the pipeline proper.
    * This problem is currently being tracked here: https://github.com/nf-core/nanoseq/issues/141

container override example:
```
process {
    withName: NANOPLOT {
        container = 'https://depot.galaxyproject.org/singularity/nanoplot:1.32.1--py_0'
    }
}
```
Place the above lines in a file called `custom.config`. When the pipleine is run, specify a new command-line parameter `-c path/to/custom.config` to the nextflow command, where `path/to/custom.config` would be the path to the custom config file.
