<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Austyn Trull and Lara Ianov, Ph.D." />


<title>nf-core tips</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">U-BDS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="nf_core_tips.html">nf-core tips</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-cube"></span>
     
    Intro to Docker with RStudio
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="intro_to_docker_rstudio_part1.html">Part 1</a>
    </li>
    <li>
      <a href="intro_to_docker_rstudio_part2.html">Part 2</a>
    </li>
    <li>
      <a href="intro_to_docker_rstudio_part3.html">Part 3</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data wrangling examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Mouse_cortex_hippocampus_Allen_brain.html">Allen Brain Map - Mouse single-cell dataset</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">nf-core tips</h1>
<h4 class="author">Austyn Trull and Lara Ianov, Ph.D.</h4>

</div>


<div id="general-notes" class="section level1">
<h1>General Notes</h1>
<ul>
<li>Use at minimum 3 samples when running a pipeline (even for minimal
testing purposes). Some pipelines will not be able to complete
successfully when running only a single sample.</li>
<li>Although not a requirement, we advise to not use the iGenomes option
suggested in a number of pipelines due to the lack of information on
versioning linked to each assembly which may compromise reproducibility
to a small degree. Instead pass on a reference (genome, and annotation
file(s)) from a source and version that is suitable to your needs.</li>
<li>Some tips that we have found require the use of a custom config.
Custom configs allow for modifying the values the pipeline uses,
including passing in new command flags to the underlying software the
pipeline uses. Custom configs are passed in via the command line using
<code>-c ${custom_config}</code>, where <code>-c</code> is the command
flag and <code>${custom_config}</code> is the path to the custom config.
<ul>
<li>Importantly, a custom config should only reflect arguments or
modifications linked to the process if not available through the
pipeline parameters. This is linked to an issue in Nextflow which is
being tracked at <a
href="https://github.com/nextflow-io/nextflow/issues/2662"
class="uri">https://github.com/nextflow-io/nextflow/issues/2662</a>.
Thus, standard pipeline parameters should be provided through the
command line during the execution of the pipeline or in a yaml file with
the usage of the <code>-params-file</code> option (e.g.:
<code>-params-file params.yml</code> where the yaml file contains all
custom parameters). Discussion linked to a <code>params.yml</code> file
can be found at <a
href="https://github.com/nf-core/viralrecon/issues/283"
class="uri">https://github.com/nf-core/viralrecon/issues/283</a> and <a
href="https://github.com/nf-core/rnaseq/issues/754"
class="uri">https://github.com/nf-core/rnaseq/issues/754</a>.</li>
</ul></li>
<li>Results from the <code>work</code> directory are copied over to the
resulting directory (typically called <code>results</code>) - deleting
the work directory may be desired to avoid duplication of files, however
deleting this directory also removes the log files (the
<code>.command.*</code> files) and prevents you from resuming a pipeline
execution (with <code>-resume</code>). If you <strong>do not</strong>
need to resume a pipeline run, and would like to remove large files from
the <code>work</code> directory while keeping the logs, you may clean
this directory with the command <code>nextflow clean -k -f</code> (and
we recommend performing a dry-run prior to this with
<code>nextflow clean -k -n</code>).</li>
</ul>
</div>
<div id="executor-mode-for-cheaha-profile" class="section level1">
<h1>Executor mode for Cheaha profile</h1>
<ul>
<li>U-BDS has created a public Cheaha profile which has been merged to
the main nf-core configs (<a
href="https://github.com/nf-core/configs/pull/322"
class="uri">https://github.com/nf-core/configs/pull/322</a>). Thus, the
profile can be used in any nf-core pipeline with the parameter
<code>-profile cheaha</code>. Full documentation of the profile can be
found in the nf-core config GitHub: <a
href="https://github.com/nf-core/configs/blob/master/docs/cheaha.md"
class="uri">https://github.com/nf-core/configs/blob/master/docs/cheaha.md</a></li>
</ul>
<p>We recommend to use this profile when processing multiple samples in
the pipeline as it will aid in parallel processing. If your run only
contains a small amount of samples, standard <code>local</code> mode (to
a compute node) is usually sufficient. Feel free to let us know if
additional customization would benefit the community.</p>
</div>
<div id="pipeline-specific-notes" class="section level1">
<h1>Pipeline-specific Notes</h1>
<div id="nf-corernaseq" class="section level2">
<h2>nf-core/rnaseq</h2>
<p>Although a user choice, we generally recommend to enable Salmon’s
flags <code>--seqBias --gcBias</code>. For versions &gt;= 3.10, users
can enable these flags by using the following pipeline parameter:
<code>extra_salmon_quant_args: "--seqBias --gcBias"</code></p>
<p><strong>Note for versions &lt; 3.10:</strong> a custom
<code>custom.config</code> is required to enable these flags. An example
is provided below for version 3.6 (this configuration provides the flags
in the salmon run with star inputs, and with the salmon run where
quasi-mapping was performed by salmon):</p>
<pre><code>// tool params not linked to direct pipeline params. 
process {
    withName: &#39;.*:QUANTIFY_STAR_SALMON:SALMON_QUANT&#39; {
        ext.args = &quot;--seqBias --gcBias&quot;
    }

    withName: &#39;.*:QUANTIFY_SALMON:SALMON_QUANT&#39; {
        ext.args = &quot;--seqBias --gcBias&quot;
    }
}</code></pre>
<p>Note that the syntax linked to custom parameters in the
<code>custom.config</code> may differ from pipeline to pipeline or even
versions. We recommend to following similar syntax to what is present in
the <code>modules.config</code> of the pipeline of interest such as <a
href="https://github.com/nf-core/rnaseq/blob/3.6/conf/modules.config">this
one for v.3.6</a>. If you have questions about this, please feel free to
ask a question linked to it during our data science office hours.</p>
<div id="tips-linked-to-reference" class="section level3">
<h3>Tips linked to reference</h3>
<ul>
<li>The <code>transcript_fasta</code> is not a required parameter by the
pipeline. If this is not provided by the user, it will be generated by
the process called <code>RSEM_PREPAREREFERENCE_TRANSCRIPTS</code> (which
may be renamed in the near-future as this process is executed whenever
the user does not provide a transcriptome, regardless of the choice of
<code>--aligner</code>). For more information, reference <a
href="https://nfcore.slack.com/archives/CE8SSJV3N/p1641824929069500">here</a></li>
<li><strong>If the user provides a file to</strong>
<code>transcript_fasta</code> which is used at the Salmon steps, it
should be a transcriptome file generated from <code>gffread</code> as
opposed to one downloaded from GENCODE or Ensembl. Issues linked to
Ensembl transcriptomes are derieved from the fact that the transcriptome
files (even after coding and non-coding RNA concatenation) have less
transcripts than the GTF file. The issues linked to GENCODE files are
linked to an issue in the pipeline which prevents it from correctly
processing the off-the-shelf transcript fasta due to the sequence name
format (despite enabling the <code>--gencode</code> parameter). For more
information, reference <a
href="https://nfcore.slack.com/archives/CE8SSJV3N/p1642697218225700">here</a>
and <a
href="https://nfcore.slack.com/archives/CE8SSJV3N/p1642699049230500">here</a>
<ul>
<li>It should be noted the current recommended way to use reference
files generally is to provide the <code>--fasta</code> and
<code>--gtf</code> along with <code>--save_reference</code> the first
time you run the pipeline and then let it generate all of the downstream
artifacts like the transcriptome and indices to be recycled thereafter
via including them in them as params in the yml. This is the way
recommended by the authors per the nf-core/rnaseq slack channel. For
more information, reference <a
href="https://nfcore.slack.com/archives/CE8SSJV3N/p1642698848229000">here</a></li>
</ul></li>
</ul>
<p><code>gffread</code> example:</p>
<pre><code>samtools faidx Mus_musculus.GRCm38.dna.primary_assembly.fa

gffread -w Mus_musculus.GRCm38_GTF_matched_transcripts.fa -g ./Mus_musculus.GRCm38.dna.primary_assembly.fa Mus_musculus.GRCm38.102.gtf</code></pre>
<p>From the example above, the newly generated transcriptome
<code>Mus_musculus.GRCm38_GTF_matched_transcripts.fa</code> would be
passed on to the pipeline.</p>
<p>We recommend to creata a conda environment with both dependencies and
record the versions.</p>
<p><code>gffread</code> documentation can be found from: <a
href="http://ccb.jhu.edu/software/stringtie/gff.shtml#gffread"
class="uri">http://ccb.jhu.edu/software/stringtie/gff.shtml#gffread</a></p>
<ul>
<li>An issue has been opened with the nf-core team to add a warning
regarding the points above linked to <code>transcript_fasta</code>: <a
href="https://github.com/nf-core/rnaseq/issues/753"
class="uri">https://github.com/nf-core/rnaseq/issues/753</a></li>
</ul>
</div>
</div>
<div id="nf-corenanoseq" class="section level2">
<h2>nf-core/nanoseq</h2>
<ul>
<li>The version of nanoplot that is used by this pipeline (1.38) causes
the pipeline to crash. Per the author, the issue is with a package that
is used to convert html files into png files. However, nanoseq expects
there to be png files which causes crash. In order to get around this
issue, you can create a custom config and override the container that
nanoseq would pull down for Nanoplot. Values specified in configs passed
via the commandline take priority over the defaults that are specified
in the pipeline proper.
<ul>
<li>This problem is currently being tracked here: <a
href="https://github.com/nf-core/nanoseq/issues/141"
class="uri">https://github.com/nf-core/nanoseq/issues/141</a></li>
</ul></li>
</ul>
<p>container override example:</p>
<pre><code>process {
    withName: NANOPLOT {
        container = &#39;https://depot.galaxyproject.org/singularity/nanoplot:1.32.1--py_0&#39;
    }
}</code></pre>
<p>Place the above lines in a file called <code>custom.config</code>.
When the pipleine is run, specify a new command-line parameter
<code>-c path/to/custom.config</code> to the nextflow command, where
<code>path/to/custom.config</code> would be the path to the custom
config file.</p>
</div>
<div id="nf-corefetchngs" class="section level2">
<h2>nf-core/fetchngs</h2>
<ul>
<li>The input file passed in via the <code>--input</code> must end in
<code>.txt</code></li>
<li>It is possible that curl may exceed the default timeout that is set
in the pipeline. In order to increase or turn this off you will need a
to add the below code to a custom config. In order to turn off curl
timelimit, the custom config will need to contain the following
lines:</li>
</ul>
<pre><code>process {
    withName: SRA_FASTQ_FTP {
        ext.args = &#39;--retry 5  --continue-at -&#39;
    }
}</code></pre>
<ul>
<li>If curl does not work, you can set the pipeline to use sra-toolkit
instead. In order to set the pipeline to use sra-toolkit to download
data, you will need to place the following lines into a
<code>params.yml</code> file, and pass that to the pipeline by adding
<code>-params-file params.yml</code> to the nextflow run command.</li>
</ul>
<pre><code>enable_conda: true
force_sra_downloads: true</code></pre>
<ul>
<li>Prefetch has a default filesize (20GB), however files you download
may be larger than this default. In order to increase the limit, you
will want to place the following lines into a <code>custom.conf</code>
file and pass it to the pipeline by adding <code>-c custom.conf</code>
to the nextflow run command. Note that this command sets the max size to
50GB but just tweak this number to accommodate your downloads</li>
</ul>
<pre><code>process {
    withName: SRATOOLS_PREFETCH {
        ext.args = &#39;--max-size 50G&#39;
    }
}</code></pre>
</div>
<div id="nf-coreatacseq" class="section level2">
<h2>nf-core/atacseq</h2>
<ul>
<li>Extra precaution should be taken when running versions 1.2.1 and
below in a cluster environment. Macs2 uses the <code>/tmp</code> dir on
the cluster, which is apt to run out of space while the process is
running and cause the macs2 output files to be empty (<code>/tmp</code>
is by design a small file system and it would be preferable to use user
scratch or user data). Since this version of the pipeline is DSL1, there
is not a way to modify the parameters of this process to store temporary
files in a different directory to properly resolve the issue, without
modifying the source code and executing it appropriately (this goes
beyond the scope of this guide. If more information is needed, we
recommend to stop by our data science office hours). <strong>As a result
we recommend running this pipeline on a local/personal computer rather
than on the cluster for the aforementioned versions.</strong>
<ul>
<li>For reference the pipeline will fail with the below error when this
occurs. It should be noted that this error can have other causes,
however, if you are running on a cluster it is worth trying to run it in
a non-cluster environment as well to confirm the cause.</li>
</ul>
<pre><code>Error in read.table(PeakFiles[idx], sep = &quot;\t&quot;, header = FALSE) : 
  no lines available in input
Execution halted</code></pre></li>
</ul>
</div>
<div id="nf-corescrnaseq" class="section level2">
<h2>nf-core/scrnaseq</h2>
<ul>
<li>We recommend only running versions post-v2.0.0 of this pipeline.
v2.0.0 is when the pipeline converted to DSL2.0 and resolved a number of
issues we had noted occurring in previous versions of the pipeline.</li>
<li>A more general note: this pipeline does not produce any QC results,
such as fastqc results, mapping stats, single-cell stats, nor do the
multiqc reports that are generated have a much useful information. Any
qc information the user is wishing to see will need to be executed
manually separately from the pipeline.
<ul>
<li>An exception to this note is <code>alevin_qc</code>, which is a
collection of various stats and graphs about your dataset. These qc
results are only produced when using alevin as an aligner.</li>
</ul></li>
</ul>
</div>
</div>

<html>

<hr class="solid">

<div class="footer">For more information about the UAB Biological Data Science core please visit the <a href="https://www.uab.edu/cores/ircp/bds">U-BDS website</a>. If you have an enquiry about a project or grant you would like to collaborate with us please fill out the <a href="https://www.wrike.com/frontend/requestforms/index.html?token=eyJhY2NvdW50SWQiOjQ2ODk3ODQsInRhc2tGb3JtSWQiOjUxMTIyNn0JNDc3OTg5MTE1NzQxNwlkNDgyOTg3Zjc0N2YyYmE4YWE3YThiYjVmNjQ3NzM1MzYyMTM5ZDlmN2I4NTU2NWZkZDZiMWY4MDg4MTJmYzI2">enquiry form.</a></div>
</html>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
